swig/cpp/examples/rpc_example.cpp

```diff
@@ -0,0 +1,277 @@
+        break;
+    case SR_UINT64_T:
+        cout << "= " << unsigned(tree->data()->get_uint64()) << endl;
+        break;
+    case SR_INT8_T:
+        cout << "= " << tree->data()->get_int8() << endl;
+        break;
+    case SR_INT16_T:
+        cout << "= " << tree->data()->get_int16() << endl;
+        break;
+    case SR_INT32_T:
+        cout << "= " << tree->data()->get_int32() << endl;
+        break;
+    case SR_INT64_T:
+        cout << "= " << tree->data()->get_int64() << endl;
+        break;
+     case SR_IDENTITYREF_T:
+        cout << "= " << tree->data()->get_identityref() << endl;
+        break;
+    case SR_BITS_T:
+        cout << "= " << tree->data()->get_bits() << endl;
+        break;
+    case SR_BINARY_T:
+        cout << "= " << tree->data()->get_binary() << endl;
+        break;
+    default:
+        cout << "(unprintable)" << endl;
+    }
+    return;
+}
+
+void
+print_value(shared_ptr<Val> value)
+{
+    cout << value->xpath();
+    cout << " ";
+    switch (value->type()) {
+    case SR_CONTAINER_T:
+    case SR_CONTAINER_PRESENCE_T:
+        cout << "(container)" << endl;
+        break;
+    case SR_LIST_T:
+        cout << "(list instance)" << endl;
+        break;
+    case SR_STRING_T:
+        cout << "= " << value->data()->get_string() << endl;;
+        break;
+    case SR_BOOL_T:
+	if (value->data()->get_bool())
+            cout << "= true" << endl;
+	else
+            cout << "= false" << endl;
+        break;
+    case SR_ENUM_T:
+        cout << "= " << value->data()->get_enum() << endl;;
+        break;
+    case SR_UINT8_T:
+        cout << "= " << unsigned(value->data()->get_uint8()) << endl;
+        break;
+    case SR_UINT16_T:
+        cout << "= " << unsigned(value->data()->get_uint16()) << endl;
+        break;
+    case SR_UINT32_T:
+        cout << "= " << unsigned(value->data()->get_uint32()) << endl;
+        break;
+    case SR_UINT64_T:
+        cout << "= " << unsigned(value->data()->get_uint64()) << endl;
+        break;
+    case SR_INT8_T:
+        cout << "= " << value->data()->get_int8() << endl;
+        break;
+    case SR_INT16_T:
+        cout << "= " << value->data()->get_int16() << endl;
+        break;
+    case SR_INT32_T:
+        cout << "= " << value->data()->get_int32() << endl;
+        break;
+    case SR_INT64_T:
+        cout << "= " << value->data()->get_int64() << endl;
+        break;
+     case SR_IDENTITYREF_T:
+        cout << "= " << value->data()->get_identityref() << endl;
+        break;
+    case SR_BITS_T:
+        cout << "= " << value->data()->get_bits() << endl;
+        break;
+    case SR_BINARY_T:
+        cout << "= " << value->data()->get_binary() << endl;
+        break;
+    default:
+        cout << "(unprintable)" << endl;
+    }
+    return;
+}
+
+int test_rpc_cb(const char *xpath, const sr_val_t *input, const size_t input_cnt, sr_val_t **output,\
+                size_t *output_cnt, void *private_ctx) {
```

Comment:
I tried to implement this in Python but I had some trouble with premature destructor call. I suspect the problem is in the SWIG permissions.
PyObject *s = SWIG_NewPointerObj(sess, SWIGTYPE_std__shared_ptrT_Session_t, 0 | SWIG_POINTER_ OWN);
This should be possible to fix.

I had in plan, when the new python callbacks work, to implement something similar in C++. So the API is the same in C++ and Python.
Is this OK for you? Or should the C++ and python API diverge already.

We can add virtual functions too, the only problem I have with them Is that you can have only one callback per class and I wanted to have similar API for C++/python and other bindings.